pinot-core/src/main/java/org/apache/pinot/core/operator/transform/function/CaseTransformFunction.java

```diff
@@ -0,0 +1,436 @@
+          }
+          break;
+        default:
+          if (resultMetadata.getDataType() != dataType) {
+            throw new IllegalStateException(
+                String.format("Incompatible expression types in THEN Clause [%s].", resultMetadata));
+          }
+      }
+    }
+    _resultMetadata = new TransformResultMetadata(dataType, true, false);
+    return _resultMetadata;
+  }
+
+  /**
+   * Evaluate the ProjectionBlock for the WHEN statements, returns an array with the
+   * index(1 to N) of matched WHEN clause, 0 means nothing matched, so go to ELSE.
+   *
+   * @param projectionBlock
+   * @return
+   */
+  private int[] getSelectedArray(ProjectionBlock projectionBlock) {
+    if (_selectedResults == null) {
+      _selectedResults = new int[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+    } else {
+      Arrays.fill(_selectedResults, 0);
+    }
+    for (int i = 0; i < _numberWhenStatements; i++) {
+      TransformFunction transformFunction = _whenStatements.get(i);
+      int[] conditions = transformFunction.transformToIntValuesSV(projectionBlock);
+      for (int j = 0; j < conditions.length; j++) {
+        if (_selectedResults[j] == 0 && conditions[j] == 1) {
+          _selectedResults[j] = i + 1;
+        }
+      }
+    }
+    return _selectedResults;
+  }
+
+  @Override
+  public int[] transformToIntValuesSV(ProjectionBlock projectionBlock) {
+    int[] selected = getSelectedArray(projectionBlock);
+    if (_intResults == null) {
+      _intResults = new int[DocIdSetPlanNode.MAX_DOC_PER_CALL];
+    }
+    for (int i = 0; i < _elseThenStatements.size(); i++) {
+      TransformFunction transformFunction = _elseThenStatements.get(i);
+      FieldSpec.DataType dataType = transformFunction.getResultMetadata().getDataType();
+      int blockNumDocs = projectionBlock.getNumDocs();
+      int[] evalInts;
+      if (dataType == FieldSpec.DataType.INT) {
+        evalInts = transformFunction.transformToIntValuesSV(projectionBlock);
+      } else {
+        evalInts = new int[blockNumDocs];
+        switch (dataType) {
+          case LONG:
+            ArrayCopyUtils.copy(transformFunction.transformToLongValuesSV(projectionBlock), evalInts, blockNumDocs);
+            break;
+          case FLOAT:
+            ArrayCopyUtils.copy(transformFunction.transformToFloatValuesSV(projectionBlock), evalInts, blockNumDocs);
+            break;
+          case DOUBLE:
+            ArrayCopyUtils.copy(transformFunction.transformToDoubleValuesSV(projectionBlock), evalInts, blockNumDocs);
+            break;
+          case STRING:
+            ArrayCopyUtils.copy(transformFunction.transformToStringValuesSV(projectionBlock), evalInts, blockNumDocs);
+            break;
+          default:
+            throw new IllegalStateException(String
+                .format("Cannot convert result type [%s] to [INT] for transform function [%s]", dataType,
+                    transformFunction));
+        }
+      }
```

Comment:
You don't need to check data type here, but directly call `evalInts = transformFunction.transformToIntValuesSV(projectionBlock);`. The transformFunction can handle the type casting. Same for other places

Reply:
Not really, the issue is that the transform function maybe calling identifierTransformFunction, which only support one data type.

Reply:
E.g. 
```
SELECT
  SUM(
    CASE
      WHEN price > 30 THEN price* 3
      WHEN price > 20 THEN price *2
      WHEN price > 10 THEN price
      ELSE 0
    END) AS weighted_price
FROM
    myTable
```
Then else clause will result the result data type to STRING as 0 is represented as a String literal 0. meanwhile price is INT, price *2 is DOUBLE.

When you try to get price from an IdentifierTransformFunction, it will try to get a String value and then get exceptions. The only thing is to check result data type and get value then copy to the data type of computed case statement.

Reply:
`IdentifierTransformFunction` should be able to do the casting. `LiteralTransformFunction` does not support casting, so you should explicitly check whether the transform function is literal and call `LiteralTransformFunction.getLiteral()`.
Besides, a lot of combinations are not possible. E.g. when the result is `INT` type, the `_elseThenStatements` will always be `INT` type.

This can simplify the code, but is a good to have but not must.

Reply:
This is a good point:
```
Besides, a lot of combinations are not possible. E.g. when the result is INT type, the _elseThenStatements will always be INT type.
```